### ISSUE TITLE
How to turn down verbosity?

### ISSUE THREAD

**Issue by xpe** [direnv/direnv#68]
I'd rather not see log messages when I change directories and the environment is adjusted. How can I turn that off?

**Comment by avinassh** [comment]
Can anyone tell me how do I hide messages/logs? The env variable `$DIRENV_LOG_FORMAT` is already empty in my case.

**Comment by jasoncodes** [comment]
I ended up [wrapping the hook function](https://github.com/jasoncodes/dotfiles/blob/9eaee03e84051929cf0a811e7c358b98107be106/shell/config/direnv.sh#L9) to silence the output I didn’t care about.

**Comment by avinassh** [comment]
Where I should put this `direnv.sh` file? 

also, I use zshell, so I think I can copy the contents of [hook function](https://github.com/jasoncodes/dotfiles/blob/master/shell/config/direnv.sh#L4,L14) in `~/.zshrc`. Let me try that.

**Comment by jasoncodes** [comment]
Aside: Take a look at http://github.com/freshshell/fresh if you would like to avoid having to copy shell config snippets around.

**Comment by avinassh** [comment]
That looks great! I will try that. Hope it works now. 

For my earlier query, I appended following to my `~/.zshrc`. 

```shell
# for direnv (check direnv.net)
if which direnv &> /dev/null; then
  eval "$(direnv hook zsh)"

  _direnv_hook() {
    eval "$(
      (
        (
          direnv export "$SHELL" 3>&1 1>&2 2>&3 3>&- |
            egrep -v -e '^direnv: (loading|export|unloading)'
        ) 3>&1 1>&2 2>&3 3>&-
      )
    )"
  };
fi
```

But now when I cd into the directory, it throws everything on the screen:

```shell
$cd work
export DIRENV_MTIME=1434946587;export TESTME=$'avinassh@avinassh.com';export DIRENV_DIFF=$'eJyqVipQsqqu1VHKA1JKLp5Brn5h8UBKyUpJVz-0OLWoWD-xLFPfPLlYSQcm7Rvi6esKVGBoYmxiaWJmamEOlHP3DIl3DA3x8A-Kd_V19PQBygM15iUWZzgANZeUpmTm6yXn50JVOvv7-nqGhLgSUFxbCwgAAP__jmEwfQ==';export DIRENV_DIR=$'-/Users/avi/work';
```

**Comment by jasoncodes** [comment]
I investigated this and it seems like late versions of Zsh enable multiple redirections by default which broke this script. I have pushed an updated version which should fix the problem.

**Comment by zimbatm** [comment]
A variation on the _direnv_hook function using anonymous pipes. This trick should work in bash as well.
```bash
_direnv_hook() {
  eval "$(direnv export zsh 2> >( egrep -v -e '^direnv: (loading|export|unloading)' ))"
};
```

**Comment by avinassh** [comment]
@jasoncodes still didn't work :/

@zimbatm worked as expected.... 

Thank you very much guys :smile:

**Comment by jasoncodes** [comment]
@zimbatm Nice. Thanks!

**Comment by kevinjalbert** [comment]
Doing `export DIRENV_LOG_FORMAT=` solved this for me.

I had to dig around to find this issue and the correct env var. Might be worth putting this on the README?

**Comment by macgeneral** [comment]
Just to add a small visual change for zsh / oh-my-zsh instead of the text output:
at the end of ~/.zshrc
`export DIRENV_LOG_FORMAT=`
`eval "$(direnv hook zsh)"`

then I created a copy of my favorite theme (agnoster.zsh-theme) in my oh-my-zsh custom/themes folder. It already adds the current virtual environment name at the beginning of the terminal prompt.

I added the following code to the end of the function `prompt_virtualenv` (line #: ~200):

> if [[ -n "$DIRENV_DIFF" ]]; then
> &nbsp;&nbsp;prompt_segment red white "!"
> fi

which now adds a white ! on red background after the virtual env name :)

**Comment by davinkevin** [comment]
For fish users: `set -x DIRENV_LOG_FORMAT ""` 😍
I've added this into my `~/.config/fish/config.fish` file, just before the `eval (direnv hook fish)`

**Comment by will** [comment]
I have this 
```
~> cat .config/direnv/direnvrc
export DIRENV_LOG_FORMAT=""
source $(asdf which direnv_use_asdf)
```

Which got rid of almost all of the logs, however it still prints one line

```
~> cd code/myproj
direnv: export +RUBYLIB ~PATH
~> 
```

Any tips on how to get this one gone too?

**Comment by Mdsp9070** [comment]
Any suggestions to edit the fish hook? I set the `DIRENV_LOG_FORMAT` but it's suppress all output...

Only echoing 
```
direnv unloading | loading
```

would be awesome

**Comment by Elland** [comment]
@Mdsp9070 

Better late than never. Having stumbled upon this issue myself over the weekend, a lot of digging around and going so far as to look into the direnv source code (not a go developer myself), I found the following to work:

```fish
function __direnv_export_eval --on-event fish_prompt;
    begin;
        begin;
            "/run/current-system/sw/bin/direnv" export fish
        end 1>| source
    end 2>| egrep -v -e "^direnv: export"
end;
```

I've also opened a PR regarding either improving the situation or at least better documenting it #806.

Some more background info [here](https://elland.me/posts/2021-06-13-fish-script-for-direnv.html), if interested.

**Comment by Anduh** [comment]
Glad that adding `export DIRENV_LOG_FORMAT=""` to my `.zshrc` worked, but keeping the load/unload would have been nice.

Isn't the `disable_stdin`[man](https://direnv.net/man/direnv.toml.1.html#codedisablestdincode) meant to do this? I tried to create a `config.toml` but it didn't work out, as it comlained with:

```
toml: cannot load TOML value of type []interface {} into a Go boolean
```
if I wrote is as `disable_stdin = "true"` (from the direnv man), and did nothing when it was `disable_stdin = true`. Also used the filename suggested in man.

**Comment by SeerLite** [comment]
@Anduh stdin is standard **input**, so I think it's not what we're looking for :P

**Comment by alexreg** [comment]
This is my Bash solution, in case it helps anyone else.

Use this to initialise direnv in your `.bashrc`.

```bash
eval "$(direnv hook bash)"

copy_function() {
	test -n "$(declare -f "$1")" || return 
	eval "${_/$1/$2}"
}

copy_function _direnv_hook _direnv_hook__old

_direnv_hook() {
	_direnv_hook__old "$@" 2> >(egrep -v '^direnv: (export)')
}
```

Would still be nice to get official support for this however...

**Comment by tjcrone** [comment]
I added support for a DIRENV_HIDE_DIFF environment variable here: https://github.com/tjcrone/direnv/commit/39856b6c266392aac08cdc25eb7f0c88b851fd1b. I would be happy to PR if the maintainers are interested, but a better approach might be to refactor logging so that there is a DIRENV_LOG_LEVEL variable with options like "debug", "warning", "minimal", etc.. Maybe maybe not. For my purposes I was happy to just drop the diff.

**Comment by jperras** [comment]
:+1: for `DIRENV_LOG_LEVEL`, or similar.

**Comment by ls-jad-elkik** [comment]
As I didn't see this mentioned in this issue, one motivation to silence the output is the usage with prezto / powerline10k:

```
Last login: Mon Feb 21 21:53:42 on ttys003

[WARNING]: Console output during zsh initialization detected.

When using Powerlevel10k with instant prompt, console output during zsh
initialization may indicate issues.

You can:

  - Recommended: Change ~/.zshrc so that it does not perform console I/O
    after the instant prompt preamble. See the link below for details.

    * You will not see this error message again.
    * Zsh will start quickly and prompt will update smoothly.

  - Suppress this warning either by running p10k configure or by manually
    defining the following parameter:

      typeset -g POWERLEVEL9K_INSTANT_PROMPT=quiet

    * You will not see this error message again.
    * Zsh will start quickly but prompt will jump down after initialization.

  - Disable instant prompt either by running p10k configure or by manually
    defining the following parameter:

      typeset -g POWERLEVEL9K_INSTANT_PROMPT=off

    * You will not see this error message again.
    * Zsh will start slowly.

  - Do nothing.

    * You will see this error message every time you start zsh.
    * Zsh will start quickly but prompt will jump down after initialization.

For details, see:
https://github.com/romkatv/powerlevel10k/blob/master/README.md#instant-prompt

-- console output produced during zsh initialization follows --

direnv: loading ~/path-to-a-direnv-project/.envrc
direnv: export +DOCKER_HOST
```

As mentioned in the message, I can disable "instant prompt" or set `DIRENV_LOG_FORMAT=''` globally. In such a case, if we have "whether direnv is loaded" as an env var and "the direnv diff" as well - it would allow me to use it in a custom segment in my prompt.

Does that sound like a good idea/compromise?

**Comment by guychouk** [comment]
@zimbatm 's solution works perfectly, thank you.
 `DIRENV_LOG_FORMAT=''` did not seem to help.

**Comment by brewster1134** [comment]
> @zimbatm's solution works perfectly, thank you. `DIRENV_LOG_FORMAT=''` did not seem to help.

you likely just need an `export` in there

```
export DIRENV_LOG_FORMAT=
```

**Comment by alexreg** [comment]
@brewster1134 No, that doesn't suppress the `direnv: export` messages, just some other ones. The way to do that is @zimbatm's solution (or mine for Bash).

**Comment by mystix** [comment]
> > @zimbatm's solution works perfectly, thank you. `DIRENV_LOG_FORMAT=''` did not seem to help.
> 
> you likely just need an `export` in there
> 
> ```
> export DIRENV_LOG_FORMAT=
> ```

this did the trick for me in zsh

**Comment by guychouk** [comment]
> > @zimbatm's solution works perfectly, thank you. `DIRENV_LOG_FORMAT=''` did not seem to help.
> 
> you likely just need an `export` in there
> 
> ```
> export DIRENV_LOG_FORMAT=
> ```

That's exactly right, for some reason I did not try it with `export` the first time around, now it works, thank you!

**Comment by ahawkins** [comment]
**tl;dr:** I prefer log levels instead of log formats.

I want to set something like `DIRENV_LOG_LEVEL=error`. The standard library already includes `log_status` and `log_error`. It seems possible to have some sort of log level variable as well.

I don't want the diagnostic information printed to my shell. It's noise when opening a lot of shell. I can live with the `export +PATH` lines. Those are helpful. This tells me things are working as expected. The problem is that by setting `DIRENV_LOG_FORMAT=` that also silences error messages. I spent two hours today debugging an error with the `asdf` plugin because errors were unintentionally silenced. My workaround, for the time being, is to keep `export DIRENV_LOG_FORMAT=` in `direnvrc` but add `env_strict` so I will see errors in my terminal regardless of `DIRENV_LOG_FORMAT`.

**Comment by aservedio** [comment]
Wow, 10 year old thread still opened and with activity. Impressive!

Just sharing something that I found out but doesn't seem to be mentioned in this thread and looks pretty useful so far when using many source_env commands.

`source_env some/other-file 2>/dev/null`

I still get the var names that got exported which IMO I'm fine & happy with since I like to scope and divide things, but now I can control the number of `direnv: loading something` lines that shows up! And it doesn't seem to affect the ability within those env file, to output to STDERR.

Enjoy!

**Comment by arobertn** [comment]
I'd also like to see `DIRENV_LOG_LEVEL`, in my case because I'd like to set it to DEBUG! Direnv is just hanging when it tries to load my nix shell, despite running "nix-shell" directly on the same shell.nix comes up in seconds, and there seems to be no way to find out what it's actually trying to run. Also, it would be good to have _whatever_ solution is implemented documented somewhere such as the man page. The only info I've found on `DIRENV_LOG_FORMAT` for instance is github issues about people trying to turn it off, so no indication on what it can be set to.

**Comment by scottdotau** [comment]
> Wow, 10 year old thread still opened and with activity. Impressive!
> 
> Just sharing something that I found out but doesn't seem to be mentioned in this thread and looks pretty useful so far when using many source_env commands.
> 
> `source_env some/other-file 2>/dev/null`
> 
> I still get the var names that got exported which IMO I'm fine & happy with since I like to scope and divide things, but now I can control the number of `direnv: loading something` lines that shows up! And it doesn't seem to affect the ability within those env file, to output to STDERR.
> 
> Enjoy!

This is the best solution I've seen (for my personal use, I guess) as it will not silence errors like the `DIRENV_LOG_FORMAT=`

For example:
```
strict_env
source_env .zshenv 2>/dev/null
use asdf 2>/dev/null

env_vars_required XDG_CONFIG_HOME \
  NOT_A_REAL_VAR_YA_SILLYBILLY
```

Will still notify of an error, where the first method won't.

Thx @aservedio

**Comment by Hubro** [comment]
Is there currently any way to only log errors?

I've recently started using nix-direnv, which unfortunately results in a ton of noise when I enter a project directory. Because of this, I've set `DIRENV_LOG_FORMAT` to an empty string and configured Starship to remind me that I'm using an .envrc file:

```toml
[custom.direnv]
format = "[\\[direnv\\]]($style) "
style = "fg:yellow dimmed"
when = "env | grep -E '^DIRENV_FILE='"
```

Result:

<img width="649" alt="image" src="https://user-images.githubusercontent.com/597206/210171542-08b4af01-658d-4198-8258-6da5f639a833.png">

This is fine, but I'm a little worried what happens if I have any errors in my `.envrc` file. How will I know anything is wrong? I would feel a lot better if I could silence *only* regular log messages rather than everything.

I also really like the "loading" / "unloading" messages, but I have no idea how to *only* keep those, and also display error messages. I tried one of the shell hook solutions earlier in this thread, but it crashed horribly and I gave up. I really wish direnv had more customisation when it comes to logging.

**Comment by dionysius** [comment]
For `bash` the setting **`export DIRENV_LOG_FORMAT=` is perfect to my liking**. Will still get an error if I change it (reminds me to reload), but is silent when navigating:

```
~/afolder/dionysius$ echo "export something=new" >> .envrc 
direnv: error ~/afolder/dionysius/.envrc is blocked. Run `direnv allow` to approve its content
~/afolder/dionysius$ direnv allow .
~/afolder/dionysius$ echo $something
new
~/afolder/dionysius$ cd ..
~/afolder$ echo $something

~/afolder$ cd dionysius/
~/afolder/dionysius$ echo $something
new
```

**Comment by OleMussmann** [comment]
Independent of the log messages, I would also like to be able to suppress the exported env vars. For some projects it fills basically the entire terminal. That's way too much noise.

`$ direnv allow`
`direnv: loading ~/code/python_env/.envrc`
`direnv: export +AR +AS +CC +CONFIG_SHELL +CXX +DETERMINISTIC_BUILD +HOST_PATH +IN_NIX_SHELL +LD +LD_LIBRARY_PATH +NIX_BINTOOLS +NIX_BINTOOLS_WRAPPER_TARGET_HOST_x86_64_unknown_linux_gnu +NIX_BUILD_CORES +NIX_CC +NIX_CC_WRAPPER_TARGET_HOST_x86_64_unknown_linux_gnu +NIX_CFLAGS_COMPILE +NIX_ENFORCE_NO_NATIVE +NIX_HARDENING_ENABLE +NIX_INDENT_MAKE +NIX_LDFLAGS +NIX_STORE +NM +OBJCOPY +OBJDUMP +PIPENV_ACTIVE +PIPENV_VENV_IN_PROJECT +PYTHONHASHSEED +PYTHONNOUSERSITE +PYTHONPATH +RANLIB +READELF +SIZE +SOURCE_DATE_EPOCH +STRINGS +STRIP +VIRTUAL_ENV +XML_CATALOG_FILES +_PYTHON_HOST_PLATFORM +_PYTHON_SYSCONFIGDATA_NAME +buildInputs +buildPhase +builder +configureFlags +depsBuildBuild +depsBuildBuildPropagated +depsBuildTarget +depsBuildTargetPropagated +depsHostHost +depsHostHostPropagated +depsTargetTarget +depsTargetTargetPropagated +doCheck +doInstallCheck +dontAddDisableDepTrack +libPath +name +nativeBuildInputs +out +outputs +patches +phases +propagatedBuildInputs +propagatedNativeBuildInputs +shell +shellHook +stdenv +strictDeps +system ~PATH ~QT_PLUGIN_PATH ~XDG_DATA_DIRS`

**Comment by Hubro** [comment]
I still think there should be a setting to reduce the noise from the `direnv: export ...` line. I like the log messages I get when loading direnv, it reminds me what's in there.

Really, my only issue is after I started using nix-shell with the nix-direnv plugin. Take this example of a wordpress theme directory:

```
➜ z client wordpress
direnv: loading ~/Dropbox/Projects/my-company/my-client/some-wordpress-theme/.envrc
direnv: loading https://raw.githubusercontent.com/nix-community/nix-direnv/2.2.0/direnvrc (sha256-5EwyKnkJNQeXrRkYbwwRBcXbibosCJqyIUuz9Xq+LRc=)
direnv: using nix
direnv: nix-direnv: using cached dev shell
direnv: export +AR +AS +CC +CONFIG_SHELL +CXX +MY_PROJECT_SCP_PASSWORD +MY_PROJECT_THEME_MOUNT_PATH +HOST_PATH
+IN_NIX_SHELL +LD +NIX_BINTOOLS +NIX_BINTOOLS_WRAPPER_TARGET_HOST_x86_64_unknown_linux_gnu +NIX_BUILD_CORES
+NIX_CC +NIX_CC_WRAPPER_TARGET_HOST_x86_64_unknown_linux_gnu +NIX_CFLAGS_COMPILE +NIX_ENFORCE_NO_NATIVE
+NIX_HARDENING_ENABLE +NIX_INDENT_MAKE +NIX_LDFLAGS +NIX_STORE +NM +NODE_PATH +OBJCOPY +OBJDUMP +RANLIB
+READELF +SIZE +SOURCE_DATE_EPOCH +STRINGS +STRIP +XDG_DATA_DIRS +buildInputs +buildPhase +builder +cmakeFlags
+configureFlags +depsBuildBuild +depsBuildBuildPropagated +depsBuildTarget +depsBuildTargetPropagated
+depsHostHost +depsHostHostPropagated +depsTargetTarget +depsTargetTargetPropagated +doCheck +doInstallCheck
+dontAddDisableDepTrack +mesonFlags +name +nativeBuildInputs +out +outputs +patches +phases
+propagatedBuildInputs +propagatedNativeBuildInputs +shell +shellHook +stdenv +strictDeps +system ~PATH
```

There is absolutely 0 value in this block of environment variables, it's purely noise.

It would be so much nicer if I could change that log line to something like:

```
direnv: exporting 43 variables
```

Or even better, if I could blacklist variables I don't give a crap about, so the output could be:

```
direnv: export +MY_PROJECT_SCP_PASSWORD +MY_PROJECT_THEME_MOUNT_PATH +IN_NIX_SHELL ~PATH (+38 more)
```

That would actually restore the value of this log line.

I assume this is a pretty niche use case though, so I don't expect Direnv to implement this, but it would be nice if there was a hook or something so people could make this kind of change themselves.

If there already is, *please* link me to some docs, I would love to implement this :smile:

**Comment by harrysbaraini** [comment]
For me, adding a echo to .envrc worked nicely:

_~/.config/fish/config.fish`

```
set -gx DIRENV_LOG_FORMAT ""
direnv hook fish | source
```

.envrc
```
echo "Loading direnv"

...
```

**Comment by viperML** [comment]
> I've recently started using nix-direnv, which unfortunately results in a ton of noise when I enter a project directory. Because of this, I've set `DIRENV_LOG_FORMAT` to an empty string and configured Starship to remind me that I'm using an .envrc file:
> 
> ```toml
> [custom.direnv]
> format = "[\\[direnv\\]]($style) "
> style = "fg:yellow dimmed"
> when = "env | grep -E '^DIRENV_FILE='"
> ```

Just wanted to tap into this comment, the condition can be reduced to:
```toml
when = "printenv DIRENV_FILE"
```

**Comment by alexreg** [comment]
@radekh Thanks for drawing my attention to that. The Bash semantics are a bit unfortunate, but indeed, this is how process substitutions work. I suspect your solution with `wait` will do just fine, though I was slightly concerned about long-running background jobs in the shell which could make `wait` suspend indefinitely. Therefore, I tried another fix which doesn't use process substitution whatsoever. Just plain old piping.

```
_direnv_hook() {
	_direnv_hook__old "$@" 2>&1 | grep -E -v '^direnv: (export)'
	true
}
```

(The rest of the code remains the same.)

**Comment by Arian04** [comment]
> I still think there should be a setting to reduce the noise from the `direnv: export ...` line. I like the log messages I get when loading direnv, it reminds me what's in there.
> 
> ...

> ```sh
> direnv: exporting 43 variables
> ```
> 
> Or even better ... the output could be:
> 
> ```sh
> direnv: export +MY_PROJECT_SCP_PASSWORD +MY_PROJECT_THEME_MOUNT_PATH +IN_NIX_SHELL ~PATH (+38 more)
> ```

I spent way too much time with awk doing something similar. Except right now it just lists the number of vars exported if it's too long

Heads up, I use zsh and haven't tested this with bash but I assume it'll work fine
```zsh
# stolen from user "alexreg" above in this issue 
# https://github.com/direnv/direnv/issues/68#issuecomment-1003426550
copy_function() {
  test -n "$(declare -f "$1")" || return
  eval "${_/$1/$2}"
}
copy_function _direnv_hook _direnv_hook__old
_direnv_hook() {
  # old line
  #_direnv_hook__old "$@" 2> >(grep -E -v '^direnv: (export)')

  # my new line
  _direnv_hook__old "$@" 2> >(awk '{if (length >= 200) { sub("^direnv: export.*","direnv: export "NF" environment variables")}}1')

  # as suggested by user "radekh" above
  wait
}
```

you're probably gonna want to adjust `length >= 200` to your liking

**Comment by lc-guy** [comment]
@Arian04 Replacing the `wait` call with `wait $!` is preferable in versions of bash >4.4 to make sure it is waiting only on the process substitution, and not every other background process running in the current shell session.

**Comment by oryband** [comment]
@Arian04 i get `direnv: error exit status 1` when trying your example, any idea why?

**Comment by Arian04** [comment]
no idea why that would be happening if your direnv doesn't through errors without my hook added. does my commented out "old line" work if you use that instead? also are you using bash or zsh?

**Comment by cheater** [comment]
i've been trying to read through all of this, but i can't get through it. would someone mind summarizing what to do if i just want the message "direnv loaded" to show up and no other spam, when i enter a direnv enabled directory or "direnv unloaded" when i leave it?

similarly, what if someone wants to have the word "direnv" to show up in their prompt in zsh or bash?

this is the top hit on google so i'm sure this will help others.

**Comment by jankatins** [comment]
To remove the "export" message completely, add `hide_env_diff = true` to `~/.config/direnv/direnv.toml` in the `global` section:

```toml
[global]
# https://direnv.net/man/direnv.toml.1.html
hide_env_diff = true
```

Watch https://github.com/direnv/direnv/pull/1231 to see if there will be a "export x environment variables" option.

**Comment by tspearconquest** [comment]
@janko wrote:
> My proposition is just to remove the difference in environment variables. So, currently it looks like this:
> 
> $ cd project
> direnv: loading .envrc
> direnv export: ~PATH, +AWS_AMAZON_KEY_ID, +AWS_AMAZON_KEY_SECRET
> 
> $ cd ..
> direnv: unloading
> direnv export: ~PATH, -AWS_AMAZON_KEY_ID, -AWS_AMAZON_KEY_SECRET
> But a user actually only needs to be reminded that .envrc was loaded/unloaded when he `cd`s in/out:
> 
> $ cd project
> direnv: loading .envrc
> 
> $ cd ..
> direnv: unloading

I believe #1233 was intended to fix this specific call out but I would like to mention that it fails to account for this line:
> But a user actually only needs to be reminded that .envrc was loaded/unloaded when he `cd`s in/out

Specifically, the "when he `cd`s in/out" part is NOT yet addressed. The `hide_env_diff` config setting hides the env diff when a shell is started in a directory with a .envrc file from an IDE like vscode, where a specific `cd` command in was not really being run; more like the initial directory was set to that as a shell argument.

I encounter giant warning messages from my zsh theme powerlevel10k when my terminal app reopens after I've restarted my Mac because I was already in a directory where direnv loads the .envrc before the restart, and zsh is starting up fresh in that directory.

In my opinion, a good fix would validate whether direnv is being called as part of shell startup scripts, and if it IS being called as part of shell startup scripts, then initialization of direnv should be silent. This way, when a user cd's in as janko mentioned, the loading message will appear, but if a user spawns a shell already in a directory where direnv will load variables, the loading will be silent.

**Comment by cheater** [comment]
> In my opinion, a good fix would validate whether direnv is being called as part of shell startup scripts, and if it IS being called as part of shell startup scripts, then initialization of direnv should be silent. This way, when a user cd's in as janko mentioned, the loading message will appear, but if a user spawns a shell already in a directory where direnv will load variables, the loading will be silent.

This is a kind of antipattern where devs trying to make a script too smart ends up backfiring on the user. Instead provide an env flag (eg `export DIRENV_NO_LOADING_MSG=1`) or a command line option, and let people control it on their own.

**Comment by tspearconquest** [comment]
Hi, thanks for mentioning this. I'm not asking for a change to direnv but rather to the shell function that calls direnv (which itself is output by `direnv hook zsh`)

```
_direnv_hook() {
  trap -- '' SIGINT
  eval "$("/opt/homebrew/bin/direnv" export zsh)"
  trap - SIGINT
}
```

This is a zsh specific hook already in direnv, therefore not really an antipattern to ask for a tweak to the shell hook to determine whether the hook is being run only when someone `cd`s into a directory versus when the shell starts up.

My thinking here is that if my theme can detect that zsh is starting up and emit a message like:

> [WARNING]: Console output during zsh initialization detected.

Then direnv can also detect that as part of this hook. Therefore no issue, and no antipattern.

**Comment by tspearconquest** [comment]
I did some testing as I wasn't sure how `direnv export zsh` works since it outputs nothing in my normal environment, so I started up a bash shell with no environment: `env -i bash` then exported 3 variables to get direnv working:

```
XDG_DATA_HOME=~/.local/share/
XDG_CACHE_HOME=~/.cache
DIRENV_CONFIG=~/.config/direnv/direnv.toml
```

Then ran `direnv export zsh` and saw what it does, which is to output the warning message and the exports for the directory you're in based on the .envrc file. Therefore I think that tweaking the zsh hook should be the right way to fix it. I'll look to see how powerlevel10k detects zsh initialization and have a go at copying that to this hook this upcoming week sometime.

**Comment by tspearconquest** [comment]
Looking at the `direnv hook zsh` output a little further, there are these lines:

```
typeset -ag precmd_functions
if (( ! ${precmd_functions[(I)_direnv_hook]} )); then
  precmd_functions=(_direnv_hook $precmd_functions)
fi
typeset -ag chpwd_functions
if (( ! ${chpwd_functions[(I)_direnv_hook]} )); then
  chpwd_functions=(_direnv_hook $chpwd_functions)
fi
```

I looked into these, was not familar with this part of zsh before. Apparently `precmd_functions` is an array of functions which zsh will run before every prompt and `chpwd_functions` is an array of functions which zsh will run when the working directory is changed.

I don't have any idea why direnv would need to be a part of `precmd_functions` so I removed that from my shell and voila - no more giant error on shell startup, and direnv still loads properly when I change directories.

A little more testing, and I figured out the `precmd_functions` is only necessary for direnv for the specific use-case of terminal startup in a directory with `.envrc` as far as I can tell. If I start up a shell from finder in a directory where `.envrc` exists, that's exactly what I don't want to happen, so this solves it completely for me, and once my prompt loads, I can just `cd .` to make it load the env.

No changes are necessary to direnv -- for users who want to avoid the issue I described, just simply comment out these 4 lines from the direnv hook wherever you find them:

```
typeset -ag precmd_functions
if (( ! ${precmd_functions[(I)_direnv_hook]} )); then
  precmd_functions=(_direnv_hook $precmd_functions)
fi
```

**Comment by injust** [comment]
Looks like an empty `DIRENV_LOG_FORMAT` no longer works in direnv v2.36.0.

**Comment by max-legrand** [comment]
For anyone on v2.36.0, creating a file at `~/.config/direnv/direnv.toml` with the following contents has helped to silence the log output for me:
```
[global]
log_format = "-"
log_filter = "^$"
```

**Comment by injust** [comment]
@max-legrand What does the `log_filter` regex do? I added only `log_format = '-'` to my config and it seems to silence the logging.

**Comment by max-legrand** [comment]
The regex should filter out all logs. When I had set only the format I was still seeing some logs. This might be a bit overkill for some, but for suppressing all logs this worked for me.

**Comment by injust** [comment]
> Looks like an empty `$DIRENV_LOG_FORMAT` no longer works in [v2.36.0](https://github.com/direnv/direnv/releases/tag/v2.36.0).

Okay this is weird...if I create an _empty_ `direnv.toml` file, empty `$DIRENV_LOG_FORMAT` works again.

**Comment by injust** [comment]
> When I had set only the format I was still seeing some logs.

I think that's a bug, since I'm seeing the literal `-` being logged. I opened https://github.com/direnv/direnv/issues/1419 for `log_format = '-'` not working properly.

**Comment by Nitive** [comment]
To disable direnv output we use configuration now

1. Create configuration file if not exists `~/.config/direnv/direnv.toml`
2. Add `log_filter=".*"` to configuration
	```toml
	[global]
	log_filter=".*"
	```

**Comment by nsheaps** [comment]
@injust looks like its because direnv is explicitly looking for the config file before checking for the env var

https://github.com/direnv/direnv/blob/master/internal/cmd/config.go#L138

I think [this change](https://github.com/direnv/direnv/pull/1336) broke some expectations about direnv configuration, introduced a few new bugs, and documentation and versioning was not updated accordingly. Namely:

* This takes a value that was honored when set in env vars by the project's envrc (like warn timeout) and forces it into user configuration
* One of the best use cases of direnv is setting stuff up automatically for users. That's great when it's scoped per project, but if it's a user configuration then it implies it shouldn't be the same for one user across many projects. That simply isn't the case here, sometimes direnv logging needs to be louder.
* The presence of a config toml in user space shouldn't change whether or not env vars are used for config per app
* Lack of documentation for examples of "all logs"/status logs/error logs for formatting
* Lack of ability to programatically change log formatting (we have it turned off normally but can turn on with an env var when direnv reloads)
* a breaking change was introduced in a minor version update without any backwards compatibility
* Previous ansi color codes in log formats no longer work

I'd like to propose the following:

1. Restore previous functionality under a new patch version 2.36.1. That means env var configs need to work regardless of a presence of the users' config
2. Make direnv.toml work at the project level such that user < project < env vars. This makes it so projects can choose to migrate to a project level config but it continues to work for previous configs. 
  * This file should be considered for whether or not it's integrity hash matches previous, as it can drastically change how direnv works.
4. Make `direnv init` create direnv.toml with defaults, prompt for project + user separately
5. Make `direnv doctor` to diagnose setup issues (these go beyond this but are issues my users frequently encountered until automation fixed it for them)  * Has the direnv hook executed
  * Is it present in this shell's rc file
  * Do you have env vars set for direnv configs
  * Is the toml present in the project? is it valid? are any values being deprecated?
  * Same at user level
  * Warn when user config != project config != env config
  * Is direnv up to date? if not, suggest updating with <install method>

**Comment by dhood** [comment]
since this is the top google result for me, `export DIRENV_LOG_FORMAT=""` fixed it for me. 

Saying explicitly because this comment from above might discourage someone who runs `echo $DIRENV_LOG_FORMAT` and finds it empty, not realising the distinction between empty and unset:

> Can anyone tell me how do I hide messages/logs? The env variable `$DIRENV_LOG_FORMAT` is already empty in my case.

(the commenter probably knows the difference, but random user may not). thanks maintainers 👍
